include/npcomp/RefBackend/Runtime/UserAPI.h

```diff
@@ -150,6 +154,131 @@ class Tensor {
   // Sizes are tail-allocated.
 };
 
+// RuntimeValue is a generic tagged union used to hold all value types
+// The tag determines the type, and the payload represents the stored 
+// contents of an object. If an object is not trivially destructible,
+// then it must be refcounted and must have a refCount. 
+#define NPCOMP_FORALL_TAGS(_)                                                  \
+  _(None)                                                                      \
+  _(Bool)                                                                      \
+  _(Int)                                                                       \
+  _(Double)                                                                    \
+  _(Tensor)
+
+struct RuntimeValue final {
+
+  RuntimeValue() : payload{0}, tag(Tag::None) {}
+
+  // Bool
+  RuntimeValue(bool b) : tag(Tag::Bool) { payload.asBool = b; }
+  bool isBool() const { return Tag::Bool == tag; }
+  bool toBool() const {
+    assert(isBool());
+    return payload.asBool;
+  }
+
+  // Int
+  RuntimeValue(std::int64_t i) : tag(Tag::Int) { payload.asInt = i; }
+  RuntimeValue(std::int32_t i) : RuntimeValue(static_cast<int64_t>(i)) {}
+  bool isInt() const { return Tag::Int == tag; }
+  bool toInt() const {
+    assert(isInt());
+    return payload.asInt;
+  }
+
+  // Double
+  RuntimeValue(double d) : tag(Tag::Double) { payload.asDouble = d; }
+  bool isDouble() const { return Tag::Double == tag; }
+  bool toDouble() const {
+    assert(isDouble());
+    return payload.asDouble;
+  }
+
+  // Tensor
+  RuntimeValue(Ref<Tensor> tensor) : tag(Tag::Tensor) {
+    payload.asRefPtr = static_cast<void *>(tensor.takePtr());
+  }
+  bool isTensor() const { return Tag::Tensor == tag; }
+  Ref<Tensor> toTensor() const {
+    assert(isTensor());
+    return Ref<Tensor>(reinterpret_cast<Tensor *>(payload.asRefPtr));
+  }
```

Comment:
The interface around passing the `Ref<T>`'s back / forth between the Union's could use some looking over.